app/models/institution_tree.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+module InstitutionTree
+  def self.build(facility_code, version)
+    main = main_campus(facility_code, version)
+    all = descendants(main.facility_code, version)
+
+    {
+      'main' => {
+        'institution' => main,
+        'branches' => build_branches(all),
+        'extensions' => build_extensions(all, main.facility_code)
+      }
+    }
+  end
+
+  class << self
+    private
+
+    def build_branches(institutions)
+      branches = []
+      institutions.each do |institution|
+        next unless institution['campus_type'] == 'N'
+        branches << {
+          'institution' => institution,
+          'extensions' => build_extensions(institutions, institution['facility_code'])
+        }
+      end
+      branches.sort_by { |branch| branch['institution']['institution'] }
+    end
+
+    def build_extensions(institutions, branch_facility_code)
+      extensions = []
+      institutions.each do |institution|
+        next unless institution['parent_facility_code_id'] == branch_facility_code && institution['campus_type'] == 'E'
+        extensions << institution
+      end
+      extensions.sort_by { |extension| extension['institution'] }
+    end
+
+    def main_campus(facility_code, version)
```

Comment:
And down here you no longer have to look up the institution:
```suggestion
    def main_campus(institution)
```